examples/C++/DDS/BasicConfigurationExample/HelloWorldPublisher.cpp

```diff
@@ -0,0 +1,300 @@
+    return true;
+}
+
+HelloWorldPublisher::~HelloWorldPublisher()
+{
+    if (writer_ != nullptr)
+    {
+        publisher_->delete_datawriter(writer_);
+    }
+    if (publisher_ != nullptr)
+    {
+        participant_->delete_publisher(publisher_);
+    }
+    if (topic_ != nullptr)
+    {
+        participant_->delete_topic(topic_);
+    }
+    DomainParticipantFactory::get_instance()->delete_participant(participant_);
+}
+
+void HelloWorldPublisher::PubListener::on_publication_matched(
+        eprosima::fastdds::dds::DataWriter*,
+        const eprosima::fastdds::dds::PublicationMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        matched_ = info.total_count;
+        std::cout << "Publisher matched." << std::endl;
+        if (enough_matched())
+        {
+            awake();
+        }
+    }
+    else if (info.current_count_change == -1)
+    {
+        matched_ = info.total_count;
+        std::cout << "Publisher unmatched." << std::endl;
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for PublicationMatchedStatus current count change" << std::endl;
+    }
+}
+
+void HelloWorldPublisher::PubListener::set_num_wait_matched(
+        uint32_t num_wait_matched)
+{
+    num_wait_matched_ = num_wait_matched;
+}
+
+bool HelloWorldPublisher::PubListener::enough_matched()
+{
+    return matched_ >= num_wait_matched_;
+}
+
+void HelloWorldPublisher::PubListener::wait()
+{
+    std::unique_lock<std::mutex> lck(wait_matched_cv_mtx_);
+    wait_matched_cv_.wait(lck, [this]
+            {
+                return enough_matched() || is_stopped();
+            });
+}
+
+void HelloWorldPublisher::PubListener::awake()
+{
+    wait_matched_cv_.notify_one();
+}
+
+void HelloWorldPublisher::runThread(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    if (samples == 0)
+    {
+        while (!is_stopped())
+        {
+            if (listener_.enough_matched())
+            {
+                publish();
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+                std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+            }
+            else
+            {
+                listener_.wait();
+            }
+        }
+    }
+    else
+    {
+        for (uint32_t i = 0; i < samples; ++i)
+        {
+            if (is_stopped())
+            {
+                break;
+            }
+            if (listener_.enough_matched())
+            {
+                publish();
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+                std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+            }
+            else
+            {
+                --i;
+                listener_.wait();
+            }
+        }
+    }
```

Comment:
Both cases are easily joinable
```suggestion
        while (!is_stopped() && (samples == 0 || hello_.index() < samples))
        {
            if (listener_.enough_matched())
            {
                publish();
                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
                          << " SENT" << std::endl;
                std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
            }
            else
            {
                listener_.wait();
            }
        }

```

Reply:
:+1: